flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java

```diff
@@ -112,7 +112,7 @@ int selectNextInputIndex(int lastReadInputIndex) {
 	}
 
 	boolean shouldSetAvailableForAnotherInput() {
-		return availableInputsMask != allSelectedMask && inputSelection.areAllInputsSelected();
+		return (inputSelection.getInputMask() & allSelectedMask & ~availableInputsMask) != 0;
```

Comment:
Could we avoid 1 `& allSelectedMask` here? It feels like it should only be relevant for `ALL`. We could then eagerly apply `allSelectedMask` instead of setting ALL (`new InputSelection(-1 & allSelectedMask)`).

Reply:
What's the problem? I think performance wise it doesn't matter, the `& allSelectedMask` would still need to be there somewhere on the critical path. 

From logic perspective, I actually prefer keeping `ALL` as valid value for the `this.inputSelection` and doing the `& allSelectedMask` here, as those extra bits are not an issue in general, but only for this particular line of code (`!= 0` comparison).

Reply:
> What's the problem? I think performance wise it doesn't matter, the `& allSelectedMask` would still need to be there somewhere on the critical path.

The point is that it would be moved out of the critical path. You only need to do it once in the ctor.

Reply:
Unfortunately no. We would still need to do this action once per record around here:
```
org.apache.flink.streaming.runtime.io.MultipleInputSelectionHandler#nextSelection
	void nextSelection() {
		if (inputSelector == null) {
			inputSelection = InputSelection.ALL;
		} else {
			inputSelection = inputSelector.nextSelection();
		}
	}
```
? And I would doubt that saving one bit operation would be visible anywhere, while as I wrote logically I think it's clearer to have this operation here.

Reply:
Okay then keep as is.